### PR TITLE
Refactor WiFiHandler and add support for multiple WiFi credentials

### PR DESCRIPTION
**Pull request by TheDevMinerTV**
I've implemented support for multiple WiFi credentials with automatic switching between networks.

This also adds new serial commands to manage the WiFi credentials:
- `WIFI ADD "<SSID>" "<PASSWORD>"`
- `WIFI REMOVE "<SSID>"`
- `WIFI LIST`

cc: @BluScream

Closes #162

### CODE REVIEW

File: src/network/WiFiHandler.h
Comment (TheButlah) [inline src/network/WiFiHandler.h:44]: Add comment, what does this do?

File: src/network/WiFiHandler.cpp
Comment (TheButlah) [inline src/network/WiFiHandler.cpp:3]: Exclude personal copyright identifier, it sets a bad precedent where all files will need to include this. Your copyright is present regardless of whether its written in the file. You never signed a contributor license agreement.

TBH I wouldn't include the MIT header either, but thats up to @Eirenliel

File: src/configuration/WiFiCredential.h
Comment (TheButlah) [inline src/configuration/WiFiCredential.h:3]: niet

File: src/network/WiFiHandler.h
Comment (TheButlah) [inline src/network/WiFiHandler.h:3]: reee

File: src/network/WiFiHandler.h
Comment (TheButlah) [inline src/network/WiFiHandler.h:50]: huge structs should be passed as const reference.

File: src/network/network.cpp
Comment (TheButlah) [inline src/network/network.cpp:3]: "SlimeVR contributors" is not a legal entity and should not appear in a copyright header. Either put Eiren or remove it completely.

File: src/network/network.h
Comment (TheButlah) [inline src/network/network.h:3]: ree

File: src/status/Status.cpp
Comment (TheButlah) [inline src/status/Status.cpp:3]: ree

File: src/status/Status.h
Comment (TheButlah) [inline src/status/Status.h:3]: ree

File: src/serial/serialcommands.cpp
Comment (TheButlah) [inline src/serial/serialcommands.cpp:67]: magic numbers, what do they mean

File: src/serial/serialcommands.cpp
Comment (TheButlah) [inline src/serial/serialcommands.cpp:88]:
```suggestion
            for (const auto& wifiCredential : *configuration.getWiFiCredentials()) {
```

File: src/network/WiFiHandler.cpp
Comment (TheButlah) [inline src/network/WiFiHandler.cpp:221]: if it can't do this, remove implementation and comment out declaration in header, mark it as TODO

File: src/network/WiFiHandler.cpp
Comment (TheButlah) [inline src/network/WiFiHandler.cpp:35]: use `::` syntax instead of nesting

File: src/network/WiFiHandler.cpp
Comment (TheButlah) [inline src/network/WiFiHandler.cpp:155]: don't do c-style casts, use `static_cast` or `reinterpret_cast`

File: src/network/network.h
Comment (TheButlah) [inline src/network/network.h:33]: is the pointer const or is the sensor const? I believe this makes the pointer const but not the underlying `Sensor`. Is that intended?

Comment (TheButlah) [inline src/network/network.h:33]: Why is it using pointers instead of references?

File: src/configuration/Configuration.h
Comment (TheButlah) [inline src/configuration/Configuration.h:34]: use `::` instead of nesting

File: src/configuration/Configuration.h
Comment (TheButlah) [inline src/configuration/Configuration.h:54]: opinion: we should start using c++17 `std::string_view`

File: src/configuration/Configuration.h
Comment (TheButlah) [inline src/configuration/Configuration.h:52]: why not `unordered_map`?

File: src/configuration/Configuration.h
Comment (TheButlah) [inline src/configuration/Configuration.h:66]: why not `unordered_map`

File: src/configuration/Configuration.cpp
Comment (TheButlah) [inline src/configuration/Configuration.cpp:438]:
```suggestion
            for (const auto& c : m_WiFiCredentials) {
```

File: src/configuration/Configuration.cpp
Comment (TheButlah) [inline src/configuration/Configuration.cpp:329]:
```suggestion
                    f.read(reinterpret_cast<uint8_t*>(&wifiCredential), sizeof(WiFiCredential));
```

File: src/configuration/Configuration.cpp
Comment (TheButlah) [inline src/configuration/Configuration.cpp:361]:
```suggestion
                    f.read(reinterpret_cast<uint8_t*>(&wifiCredential), sizeof(WiFiCredential));
```

File: src/configuration/Configuration.cpp
Comment (TheButlah) [inline src/configuration/Configuration.cpp:274]:
```suggestion
                    f.read(reinterpret_cast<uint8_t*>(&calibrationConfig), sizeof(CalibrationConfig));
```

**Review by TheButlah — CHANGES_REQUESTED** [review]
I didn't test the code or review the actual functionality, most of my comments were style nits.

Please check with @Eirenliel about the member variable naming conventions and `this->` to determine what was decided there.

**Comment by Kamilake** [comment]
I think support for legacy Wi-Fi connections should be left behind for a while
Many people, including myself, use the old client. so i think when using deprecated commands, that must print out warnings and connect to Wi-Fi.

File: src/serial/serialcommands.cpp
Comment (Bluscream) [reply src/serial/serialcommands.cpp:67]: anna already complained about too many comments in another PR :joy:

File: src/configuration/WiFiCredential.h
Comment (TheDevMinerTV) [reply src/configuration/WiFiCredential.h:3]: look at the file, git thought I copied it, but I didn't, I entirely rewrote it.

File: src/network/WiFiHandler.h
Comment (TheDevMinerTV) [reply src/network/WiFiHandler.h:43]: it's `setup()` because it's called from `setup()`, and at that point the other stuff has also been initialized.

File: src/GlobalVars.h
Comment (TheDevMinerTV) [reply src/GlobalVars.h:27]: but if you include it in `LEDManager.h` or anything that's here, it won't compile.

File: src/configuration/Configuration.cpp
Comment (TheDevMinerTV) [reply src/configuration/Configuration.cpp:49]: I just wanna get the file looking uniform, atleast this one, the new code is written with `this->`.

File: src/network/WiFiHandler.h
Comment (TheDevMinerTV) [reply src/network/WiFiHandler.h:39]: doesn't compile on esp32

File: src/configuration/Configuration.h
Comment (TheDevMinerTV) [reply src/configuration/Configuration.h:54]: suggestion: fix up the esp32 toolchain 🙃

File: src/network/WiFiHandler.cpp
Comment (TheDevMinerTV) [reply src/network/WiFiHandler.cpp:35]: doesn't compile on esp32

File: src/configuration/Configuration.h
Comment (TheDevMinerTV) [reply src/configuration/Configuration.h:34]: doesn't compile on esp32

File: src/network/network.h
Comment (TheDevMinerTV) [reply src/network/network.h:33]: @Eirenliel

Comment (TheDevMinerTV) [reply src/network/network.h:33]: @Eirenliel

File: src/configuration/Configuration.cpp
Comment (TheDevMinerTV) [reply src/configuration/Configuration.cpp:136]: Let's not change stuff I haven't changed, you're free to open a new PR if you wanna change this

File: src/configuration/Configuration.cpp
Comment (TheDevMinerTV) [reply src/configuration/Configuration.cpp:274]: Same thing here

File: src/configuration/Configuration.h
Comment (TheDevMinerTV) [reply src/configuration/Configuration.h:66]: doesn't compile on esp32

File: src/configuration/Configuration.h
Comment (TheDevMinerTV) [reply src/configuration/Configuration.h:52]: doesn't compile on esp32

File: src/configuration/Configuration.cpp
Comment (TheDevMinerTV) [reply src/configuration/Configuration.cpp]: doesn't compile on esp32

**Comment by TheDevMinerTV** [comment]
> I think support for legacy Wi-Fi connections should be left behind for a while Many people, including myself, use the old client. so i think when using deprecated commands, that must print out warnings and connect to Wi-Fi.

Done, this is back!

**Comment by TheDevMinerTV** [comment]
Thanks to @unlogisch04 for helping me debug this stuff.
There's a couple bugs relating to the ICM but that is stuff for it's own PR.
This should be mostly usable, so I believe y'all should do a functional QA. :)

**Review by unlogisch04 — CHANGES_REQUESTED** [review]
Name Filtering for correct Sensor Calibration Data.
Maybe this should belong to a other pull request.
```
                    const char *namefile = f.name();
                    char *endfile = strrchr(namefile, '/');
                    uint8_t sensorId=0;
                    if (endfile != NULL) {
                        sensorId = strtoul(endfile+1, NULL, 0);
                    } else {
                        sensorId = strtoul(namefile, NULL, 0);
                    }
```

**Comment by unlogisch04** [comment]
There are some important drawbacks:
- Does not work with hidden SSID
- Does trigger Timeouts while searching for WiFi on I2C Sensor (icm20948, but i think others need to be checked too)
- Needs a fix regarding IMU config read for ESP32
- WIFI ADD "SSID" "PASSWORD" does not correctly save or restore after a reboot
- WIFI REMOVE "SSID" seems not to work too

**Comment by TheButlah** [comment]
@TheDevMinerTV please don't resolve conversations, the reviewer should do that. I now have to go through each one, and make sure manually that the changes were adressed :(. Makes incremental reviews impossible

File: src/GlobalVars.h
Comment (TheButlah) [reply src/GlobalVars.h:27]: I'll take your word for it

File: src/configuration/Configuration.cpp
Comment (TheButlah) [reply src/configuration/Configuration.cpp:49]: sounds good, resolving

File: src/configuration/Configuration.cpp
Comment (TheButlah) [reply src/configuration/Configuration.cpp:136]: ok, fair enough

File: src/configuration/WiFiCredential.h
Comment (TheButlah) [reply src/configuration/WiFiCredential.h:36]: You could alternatively treat the file path as containing the version info, and a different filepath would be a different version

File: src/network/WiFiHandler.h
Comment (TheButlah) [reply src/network/WiFiHandler.h:43]: when you say "the other stuff has also been initialized" can you clarify what other stuff you're referring to? Do you mean  the struct itself is already initialized (that would happen if this is a global)

File: src/configuration/WiFiCredential.h
Comment (TheButlah) [reply src/configuration/WiFiCredential.h:3]: Yeah but its still based on Eiren's copyright because the original was written by Eiren.

File: src/network/WiFiHandler.h
Comment (TheButlah) [reply src/network/WiFiHandler.h:39]: you're right, we need to upgrade `xtensa-toolchain` to get a newer gcc that supports C++17. I'll go ahead and make that change in a separate PR. That will probably resolve most of my nits in this PR. 

Feel free to mark as resolved when you've read.

File: src/configuration/Configuration.h
Comment (TheButlah) [reply src/configuration/Configuration.h:52]: I just tested this on both `esp12e` and `esp32`, both compile. `unordered_map` was added in C++ 11 so it should work fine.

File: src/configuration/Configuration.cpp
Comment (TheButlah) [reply src/configuration/Configuration.cpp]: I just tested on both `esp32` and `esp8266`, both work with `reinterpret_cast`

File: src/configuration/Configuration.h
Comment (TheButlah) [reply src/configuration/Configuration.h:67]: compiles for me

File: src/configuration/Configuration.cpp
Comment (TheButlah) [reply src/configuration/Configuration.cpp:274]: compiles fine on both platforms, reopening

**Comment by TheButlah** [comment]
OK @TheDevMinerTV I have gone through them now, some of them that were closed should not have been closed. Give it a look when you get the chance :)

File: src/configuration/Configuration.cpp
Comment (TheButlah) [reply src/configuration/Configuration.cpp]: looks like you already made the change, resolving.

File: src/configuration/Configuration.cpp
Comment (unlogisch04) [inline src/configuration/Configuration.cpp:188]: This seems not to insert a new element. 
 m_Logger.trace("m_WiFiCredentials size: %d", m_WiFiCredentials.size()); returns each time 1
```
[TRACE] [Configuration] Found WiFi credential file: /wifi_credentials/test1
[INFO ] [Configuration] Found up-to-date WiFi credential v1
[TRACE] [Configuration] m_WiFiCredentials size: 1
[TRACE] [Configuration] Found WiFi credential file: /wifi_credentials/test2
[INFO ] [Configuration] Found up-to-date WiFi credential v1
[TRACE] [Configuration] m_WiFiCredentials size: 1
[TRACE] [Configuration] Found WiFi credential file: /wifi_credentials/test3
[INFO ] [Configuration] Found up-to-date WiFi credential v1
[TRACE] [Configuration] m_WiFiCredentials size: 1
[TRACE] [Configuration] Found WiFi credential file: /wifi_credentials/test4
[INFO ] [Configuration] Found up-to-date WiFi credential v1
[TRACE] [Configuration] m_WiFiCredentials size: 1
------
[TRACE] [StatusManager] Added status WIFI_CONNECTING
[TRACE] [WiFiHandler] Adding WiFi credential: test4
[INFO ] [WiFiHandler] Loaded credentials for "test4"
```

File: src/network/network.cpp
Comment (TheButlah) [reply src/network/network.cpp:3]: Resolving, as I just don't care enough, and because who really does?
I'm also resolving my other nits about this topic, those can be brought up in a different PR

File: src/configuration/WiFiCredential.h
Comment (TheButlah) [reply src/configuration/WiFiCredential.h:3]: Resolving things related to this, those can be brought up in a different PR. This PR review is lasting too long.

Resolve when you've read

File: src/configuration/Configuration.cpp
Comment (TheButlah) [reply src/configuration/Configuration.cpp:188]: This is probably why: https://stackoverflow.com/a/4157729

Comment (TheButlah) [reply src/configuration/Configuration.cpp:188]: Confirmed, all the credential ssids have the same memory address, so they all have the same key.

Comment (unlogisch04) [reply src/configuration/Configuration.cpp:188]: m_Logger.trace("m_WiFiCredentials size: %d, SSID: %s, %p", m_WiFiCredentials.size(), credential.ssid, &credential.ssid);
```
[TRACE] [Configuration] Found WiFi credential file: /wifi_credentials/test1
[INFO ] [Configuration] Found up-to-date WiFi credential v1
[TRACE] [Configuration] m_WiFiCredentials size: 1, SSID: fentozell, 0x3ffb1eb8
[TRACE] [Configuration] Found WiFi credential file: /wifi_credentials/test2
[INFO ] [Configuration] Found up-to-date WiFi credential v1
[TRACE] [Configuration] m_WiFiCredentials size: 1, SSID: test2, 0x3ffb1eb8
[TRACE] [Configuration] Found WiFi credential file: /wifi_credentials/test3
[INFO ] [Configuration] Found up-to-date WiFi credential v1
[TRACE] [Configuration] m_WiFiCredentials size: 1, SSID: test3, 0x3ffb1eb8
[TRACE] [Configuration] Found WiFi credential file: /wifi_credentials/test4
[INFO ] [Configuration] Found up-to-date WiFi credential v1
[TRACE] [Configuration] m_WiFiCredentials size: 1, SSID: test4, 0x3ffb1eb8
```

File: src/configuration/WiFiCredential.h
Comment (TheDevMinerTV) [reply src/configuration/WiFiCredential.h:36]: already added this

**Comment by TheDevMinerTV** [comment]
> Does trigger Timeouts while searching for WiFi on I2C Sensor (icm20948, but i think others need to be checked too)

@unlogisch04 this has always been an issue though (atleast `main` does have the same issue with my ICMs)

File: src/configuration/Configuration.cpp
Comment (TheDevMinerTV) [reply src/configuration/Configuration.cpp:274]: Can we please not touch I didn't play around in? that's what I meant with "Same thing here"

File: src/network/WiFiHandler.h
Comment (TheDevMinerTV) [reply src/network/WiFiHandler.h:43]: We're relying on `Serial` which get's initialized in `main.cpp:55`, so therefore we have to construct the rest of the stuff after that gets initialized. That's why this function exists. (a lot of existing code does it the same way)

File: src/network/WiFiHandler.h
Comment (TheDevMinerTV) [reply src/network/WiFiHandler.h:44]: This function gets called every iteration of `main.cpp`'s `loop()` and updates the `WiFiMulti instance.

File: src/network/WiFiHandler.h
Comment (TheButlah) [reply src/network/WiFiHandler.h:43]: Gotcha, makes sense. Thanks!

File: src/network/WiFiHandler.h
Comment (TheButlah) [inline src/network/WiFiHandler.h:44]:
```suggestion
            /// Call repeatedly to drive the `WiFiHandler`
            void update();
```

File: src/network/WiFiHandler.h
Comment (TheButlah) [reply src/network/WiFiHandler.h:44]: Makes sense, I've added a suggestion

File: src/network/network.h
Comment (TheButlah) [reply src/network/network.h:33]: Feel free to mark as resolved if Eiren doesn't get back to you before the PR is ready for merge

Comment (TheButlah) [reply src/network/network.h:33]: Feel free to mark as resolved if Eiren doesn't get back to you before the PR is ready for merge

**Comment by TheButlah** [comment]
Looks nearly there, just the `map` related stuff remains I think

**Comment by TheDevMinerTV** [comment]
Already did that

File: src/network/WiFiHandler.h
Comment (TheDevMinerTV) [reply src/network/WiFiHandler.h:47]: done

**Comment by TheButlah** [comment]
Referring to the bug in the map key

**Comment by TheDevMinerTV** [comment]
@TheDevMinerTV todo for today or tomorrow

File: src/configuration/Configuration.cpp
Comment (TheDevMinerTV) [reply src/configuration/Configuration.cpp:180]: Should be fixed, please try again

**Review by unlogisch04 — CHANGES_REQUESTED** [review]
If there is no credential or no WiFi connection the tracker cycle time is about 1.5sec.
This leads that only every 1.5sec a character is read from the serial input.
The WiFiMulti lib waits till all channels are scanned before it returns to work.
There is no async option in the WiFiMulti lib implemented. WiFiScan allow a async option.
This all is for ESP8266 i did not read into ESP32.

I work currently to replace the WiFiMulti. With similar function. Also to get it work with hidden SSID.

**Comment by TheDevMinerTV** [comment]
Closing due to issues with the library used.

**Comment by Bluscream** [comment]
> Closing due to issues with the library used.

ESPHome allow setting multiple SSIDs: https://esphome.io/components/wifi.html#connecting-to-multiple-networks

**Comment by TheDevMinerTV** [comment]
But for them it doesn't matter that every tick you have to scan the WiFi environment, for us it does.

**Pull request closed without merging**
